File: topics/procedure-decl/examples/tell-joke.c
/*
* Program: tell-joke.c
* Tell a joke using a number of procedures.
*/

void setup_joke();
{
  printf("Knock Knock.\n");
  printf("Who''s there?\n");
  printf("Boo\n");
  printf("Boo Who?\n");
}

void pause_for_dramatic_effect();
{
    printf(".. pause ..\n");
}

void punch_line();
{
  pause_for_dramatic_effect();
  printf("Don't cry its only a joke!\n");
}

void tell_joke();
{
  setup_joke();
  punch_line();
}

void ha();
{
  printf("ha");
}

void ha_ha();
{
  ha(); 
  ha();
}

void he_he();
{
  printf("he_he");
}

void ah();
{
  printf("ah");
}

void ha_hahe_he();
{
  ha_ha();
  he_he();
}

void laugh();
{
  ah();
  ha_ha();
  ha_hahe_he();
  pause_for_dramatic_effect();
  he_he(); 
}

void stop_laughing();
{
  printf("\n");
  ah();
  pause_for_dramatic_effect();
  printf("\n");
  printf("Yes, very funny :)\n");
}

void main();
{
  tell_joke();
  laugh();
  stop_laughing();
}

File: topics/procedure-decl/examples/TellJoke.pas
// Program: TellJoke.pas
// Tell a joke using a number of procedures.
program TellJoke;
uses CRT;

procedure SetupJoke();
begin
  WriteLn('Knock Knock.');
  WriteLn('Who''s there?');
  WriteLn('Boo');
  WriteLn('Boo Who?');
end;

procedure PauseForDramaticEffect();
begin
  Delay(1500);
end;

procedure PunchLine();
begin
  PauseForDramaticEffect();
  WriteLn('Don''t cry its only a joke!');
end;

procedure TellJoke();
begin
  SetupJoke();
  PunchLine();
end;

procedure Ha();
begin
  Write('ha');
end;

procedure HaHa();
begin
  Ha(); Ha();
end;

procedure HeHe();
begin
  Write('hehe');
end;

procedure Ah();
begin
  Write('ah');
end;

procedure HaHaHeHe();
begin
  HaHa();
  HeHe();
end;

procedure Laugh();
begin
  Ah();
  HaHa(); 
  HaHaHeHe();
  PauseForDramaticEffect();
  HeHe(); 
end;

procedure StopLaughing();
begin
  WriteLn();
  Ah();
  PauseForDramaticEffect();
  WriteLn();
  WriteLn('Yes, very funny :)');
end;

procedure Main();
begin
  TellJoke();
  Laugh();
  StopLaughing();
end;

begin
  Main();
end.